src/relax/op/tensor/binary.cc

```diff
@@ -105,5 +114,50 @@ StructInfo InferStructInfoBroadcast(const Call& call, const BlockBuilder& ctx) {
   }
 }
 
+StructInfo InferStructInfoBroadcast(const Call& call, const BlockBuilder& ctx) {
+  auto f_compute_out_dtype = [](const Call& call, const BlockBuilder& ctx,
+                                const TensorStructInfoNode* lhs_sinfo,
+                                const TensorStructInfoNode* rhs_sinfo) {
+    DataType output_dtype;
+    if (lhs_sinfo->IsUnknownDtype() || rhs_sinfo->IsUnknownDtype()) {
+      output_dtype = DataType::Void();
+    } else if (lhs_sinfo->dtype != rhs_sinfo->dtype) {
+      ctx->ReportFatal(Diagnostic::Error(call)
+                       << "Data types " << lhs_sinfo->dtype << " and " << rhs_sinfo->dtype
+                       << " must be equal for broadcasting operators");
+    } else {
+      output_dtype = lhs_sinfo->dtype;
+    }
+    return output_dtype;
+  };
+  return InferStructInfoBinary(call, ctx, std::move(f_compute_out_dtype));
+}
```

Comment:
Shall we move the logic of broadcasting to `InferStructInfoBroadcast`? I find it still in `InferStructInfoBinary`.

Reply:
I don’t think they duplicate. Note that `InferStructInfoBinary` takes a function that computes the output datatype accordingly. `InferStructInfoBroadcast` and `InferStructInfoCMP` uses that to specialize their own way of computing output datatype.

Reply:
Oh I understand. So `InferStructInfoBroadcast` and `InferStructInfoCMP` pass `f_compute_out_dtype` to `InferStructInfoBinary`. But my point is that the main logic of calculating broadcast shape (L83~L109) is put in `InferStructInfoBinary`. Will it be good if we change some names here?

Reply:
For example,
```
InferStructInfoBinary => InferStructInfoBroadcast
```
because no matter what output type is, it will infer shape by the rule of broadcasting. And
```
InferStructInfoBroadcast => InferStructInfoBroadcastArith
InferStructInfoCMP => InferStructInfoBroadcastCMP
```
because they will both infer shape by the rule of broadcasting,  infer type by their unique `f_compute_out_dtype`.

Reply:
Sounds good.

Reply:
Updated :-)